Slider component

rreusser opened the pull request:
This PR implements a first cut at sliders #742
cc @etpinard 

It has most of the necessary features (including hooking into the internal event dispatcher :tada:) and seems reasonably stable. It needs:

- [x] design feedback
- [x] should echo the selected value since the label may get dropped if there's not enough room
- [x] should expose more configuration options
- [ ] perhaps account for continuous sliders out of the gate so we don't work ourselves into a backwards compatibility corner
- [x] fix circular deps… orrrr raise allowed circular deps 😏
- [x] make the constants less intelligent and more flexible. Just spell them out. Maybe grab tick constants from elsewhere.
- [x] tests 😬 

Notable features:

- [x] measures text and sets major/minor tick stride to avoid label overlap
- [x] subscribes to events and updates accordingly
- [x] wooooo transitions

etpinard commented:
@rreusser can you share a few screenshots here for the design team?

etpinard commented on src/components/slider/attributes.js:
```diff
@@ -0,0 +1,203 @@
+        dflt: 'restyle',
+        role: 'info',
+        description: [
+            'Sets the Plotly method to be called on click.'
```
`s/on click/on slide` maybe?

etpinard commented on src/components/slider/attributes.js:
```diff
@@ -0,0 +1,203 @@
+
+    steps: stepsAttrs,
+
+    updateevent: {
```
:tada:

rreusser commented:
Here's the slider at full size. It can be translated, but vertical orientation is not currently implemented.

<img width="982" alt="screen shot 2016-09-28 at 10 00 27" src="https://cloud.githubusercontent.com/assets/572717/18916541/b96e9a18-8562-11e6-9da4-cf14e505304b.png">

Here it is at reduced size. The label size is calculated and it drops labels so that they don't overlap.

<img width="548" alt="screen shot 2016-09-28 at 10 00 45" src="https://cloud.githubusercontent.com/assets/572717/18916535/b607baf8-8562-11e6-945a-abb8c7cb0e14.png">

etpinard commented on src/components/slider/attributes.js:
```diff
@@ -0,0 +1,203 @@
+            'value when an event of type `updateevent` is received. If',
+            'undefined, the data argument itself is used. If a string,',
+            'that property is used, and if a string with dots, e.g.',
+            '`item.0.label`, then `data[\'item\'][0][\'label\']` will',
```
Ok. `Plotly.restyle` and `Plotly.relayout` use use `[]` to delimit array items in an attribute string e.g. `item[0].label` via [`Lib.nestedProperty`](https://github.com/plotly/plotly.js/blob/708e0ffd9ee827fd2640334a072f86605dcb2f5d/src/lib/nested_property.js#L29) .

I believe the `item.0.label` is a little more clear. But, maybe we should stick to `item[0].label` for consistency?

etpinard commented on src/components/slider/attributes.js:
```diff
@@ -0,0 +1,203 @@
+        role: 'info',
+        dflt: 'fraction',
+        description: [
+            'Determines whether this color bar\'s length',
```
Haha I see, the descriptions were pasted over from other components.

rreusser replied:
Hahaha yupppp.

etpinard commented on src/components/slider/draw.js:
```diff
@@ -0,0 +1,507 @@
+        return function(data) {
+            var value = data;
+            if(updatevalue) {
+                value = Lib.nestedProperty(data, updatevalue).get();
```
wait does `nestedProperty` work with `items.0.label` signature?

etpinard replied:
maybe I just didn't know about it.

rreusser replied:
Ah, that's a good question… It may well not. I simply assumed (hoped?) it would. Thanks for noticing.

etpinard commented on src/components/slider/draw.js:
```diff
@@ -0,0 +1,507 @@
+        Plotly[step.method](gd, args[0], args[1], args[2]).catch(function() {
+            // This is not a disaster. Some methods like `animate` reject if interrupted
+            // and *should* nicely log a warning.
+            Lib.warn('Warning: Plotly.' + step.method + ' was called and rejected.');
```
nice touch

rreusser replied:
Yeah, I think this should be factored out into a general method execution library function(s). If we use this in a few places, it'll be important for this to conform to some reasonable expectations and subtleties whose implementations we don't want to simplyl duplicate.

etpinard commented on src/plotly.js:
```diff
@@ -37,6 +37,7 @@ exports.Shapes = require('./components/shapes');
 exports.Legend = require('./components/legend');
 exports.Images = require('./components/images');
 exports.UpdateMenus = require('./components/updatemenus');
+exports.Slider = require('./components/slider');
```
let's use the plural here: `Sliders`

rreusser replied:
Mainly didn't because of `"rangeslider"`

etpinard replied:
But you see, `RangeSlider` is consistent with `layout.xaxis.rangeslider` i.e. the option container name (even though we might want to support multiple range sliders in the future).

etpinard commented on src/plots/layout_attributes.js:
```diff
@@ -183,6 +183,7 @@ module.exports = {
         'images': 'images',
-        'updatemenus': 'updatemenus'
+        'updatemenus': 'updatemenus',
+        'slider': 'slider'
```
similarly here

etpinard replied:
@rreusser this line here should be `sliders: sliders`.

Why you ask? It is to make `Plotly.PlotSchema.get()` include `layout.sliders` attributes. The `PlotSchema` module looks for the component by name [here](https://github.com/plotly/plotly.js/blob/fce36fa46bd53fa5ebe4e4892479258040783409/src/plot_api/plot_schema.js#L242) to paste nested attributes in the plot schema object.

etpinard replied:
![image](https://cloud.githubusercontent.com/assets/6675409/18962371/8ae2737c-863e-11e6-9854-7a8193b7053e.png)

you see here, `sliders` is missing.

etpinard commented on src/components/updatemenus/draw.js:
```diff
@@ -383,7 +386,7 @@ function styleOnMouseOut(item, menuOpts) {
 
 // find item dimensions (this mutates menuOpts)
-function findDimenstions(gd, menuOpts) {
+function findDimensions(gd, menuOpts) {
```
thanks!

rreusser replied:
Haha, I love finding typos and then finding the compensatory typos elsewhere…

etpinard commented on src/components/slider/index.js:
```diff
@@ -0,0 +1,20 @@
+
+exports.moduleType = 'component';
+
+exports.name = 'slider';
```
plural (to be consistent with `layout.sliders`

etpinard commented on src/components/slider/draw.js:
```diff
@@ -0,0 +1,507 @@
+    });
+}
+
+function drawTicks(sliderGroup, sliderOpts) {
```
Good enough for now i.e. no need to make the slider tick options on par with axis tick options (++ `axes.js` is most fun module to reuse).

etpinard replied:
EDIT: (++ axes.js is **not** most fun module to reuse)

etpinard commented on src/components/slider/draw.js:
```diff
@@ -0,0 +1,507 @@
+    item.on('mousedown', function() {
+        var grip = sliderGroup.select('.' + constants.gripRectClass);
+
+        d3.event.stopPropagation();
```
we might want to use a `coverSlip` in the range slider part II [here](https://github.com/plotly/plotly.js/pull/962/files#r79949182) to avoid selecting other elements on the graph on drag.

etpinard replied:
Ah but looks like `d3.event.stopPropagation(); d3.event.preventDefault();` does the trick. Nevermind.

rreusser replied:
Yeah, I wasn't 100% sure about my ad-hoc drag implementation. I feel like it ended up fairly unsurprising and reasonably robust, but I'm glad to refactor if there's helpers that manage some o this.

etpinard commented on src/components/slider/draw.js:
```diff
@@ -0,0 +1,507 @@
+
+    // Lib.setTranslate doesn't work here becasue of the transition duck-typing.
+    // It's also not necessary because there are no other transitions to preserve.
+    el.attr('transform', 'translate(' + (x - constants.gripWidth * 0.5) + ',' + 0 + ')');
```
thanks for the comment

rreusser commented:
@etpinard Added mock. The big shortcoming is hooking up to events since mocks don't work with animations.

<img width="1090" alt="screen shot 2016-09-28 at 10 47 05" src="https://cloud.githubusercontent.com/assets/572717/18918486/0021286c-8569-11e6-9018-ed143bed0c9c.png">

rreusser commented on src/components/slider/draw.js:
```diff
@@ -0,0 +1,507 @@
+
+function makeInputProxy(gd, sliderGroup, sliderOpts) {
+    sliderOpts.inputProxy = gd._fullLayout._paperdiv.selectAll('input.' + constants.inputProxyClass)
+        .data([0]);
```
This is not currently implemented. The idea is that it creates an `<input type="range">` that captures focus (can it actually do that programatically?) so that accessibility things like keyboard arrows to change the slider position actually work. Just an idea. Might need to go.

etpinard replied:
Thanks for the info

rreusser commented on src/components/slider/attributes.js:
```diff
@@ -0,0 +1,221 @@
+        ].join(' ')
+    },
+
+    transition: {
```
I kept this nested format so that it perfectly mirrors the way you'd define transition config for an animation. Otherwise it would need to be `transitionduration:` and `transitioneasing:`, which isn't bad. Open to thoughts here.

etpinard replied:
:+1: plotly.js is a big fan of nested attributes.

rreusser commented on src/components/slider/draw.js:
```diff
@@ -0,0 +1,507 @@
+    sliders.exit().remove();
+
+    // If no more sliders, clear the margisn:
+    if(sliders.exit().size()) clearPushMargins(gd);
```
When sliders disappear, it also needs to detach event subscriptions. That's very important! We should think about the proper way for components to do that in general.

etpinard replied:
Good point.

rreusser commented on src/components/slider/draw.js:
```diff
@@ -0,0 +1,507 @@
+    sliderOpts.inputAreaLength = Math.round(sliderOpts.outerLength - sliderOpts.xpad * 2);
+    sliderOpts.railInset = Math.round(Math.max(0, constants.gripWidth - constants.railWidth) * 0.5);
+    sliderOpts.stepInset = Math.round(Math.max(sliderOpts.railInset, constants.gripWidth * 0.5));
+
```
The logic in this area needs to be 30% less fancy and 40% more directly controlled by constants.

rreusser replied:
Done ✅

etpinard commented:
@jackparmer @delekru

would mind taking a look at https://github.com/plotly/plotly.js/pull/986#issuecomment-250175532 ?

rreusser commented:
@etpinard thoughts on decircularizing requires vs. upping the cirularity threshold?

etpinard commented:
> thoughts on decircularizing requires vs. upping the cirularity threshold?

Every component that calls a `Plotly` method will have a circular dependency in the current state of the `src/` folder. So, your PR should indeed add another circular dependency.

That `test-syntax` block is simply there to be aware that at some point `var MAX_ALLOWED_CIRCULAR_DEPS = 0;`.

So, go ahead and bump `MAX_ALLOWED_CIRCULAR_DEPS` !

rreusser commented:
Thanks @etpinard. Will gladly bump threshold!

![fd4a1cc69ea211681649a4149050a415](https://cloud.githubusercontent.com/assets/572717/18919780/5b43f3b0-856d-11e6-955f-dc6e36c1a2f9.gif)

jackparmer commented:
Love these sliders. 💃 from me on the design side.

etpinard commented on src/components/sliders/draw.js:
```diff
@@ -0,0 +1,532 @@
+        .data([0]);
+}
+
+// This really only just filters by visibility:
```
yep, you got it :microscope:

rreusser commented on src/components/sliders/defaults.js:
```diff
@@ -0,0 +1,120 @@
+    coerce('updatevalue');
+
+    if(!sliderIn.transition) {
+        sliderIn.transition = {};
```
No reason. I've removed it.

rreusser replied:
(Apart from me not previously having realized that `coerce` creates nested objects just fine.)

etpinard commented on src/components/sliders/draw.js:
```diff
@@ -0,0 +1,532 @@
+        width: constants.gripWidth,
+        height: constants.gripHeight,
+        rx: constants.gripRadius,
+        ry: constants.gripRadius,
```
My only design comment is: maybe we could make the grip a little more rectangular i.e. more similar to how the range slider looks:

![image](https://cloud.githubusercontent.com/assets/6675409/18962010/070a1402-863d-11e6-9ce8-27a079463751.png)


... but I'm no design expert.

delekru commented:
@rreusser @etpinard 

I really like the way this version of the slider looks:
![slider](https://cloud.githubusercontent.com/assets/11370069/18963628/cf52c412-8643-11e6-8f3f-1f240a1042ea.png)

As for the position of the current value, I think above the slider would be better than below because the labels are already displayed there. Maybe something similar to this?
![currentyear](https://cloud.githubusercontent.com/assets/11370069/18964406/bb786570-8646-11e6-9aa8-32bd8a71c380.png)

rreusser commented on src/components/sliders/attributes.js:
```diff
@@ -0,0 +1,255 @@
+                'When currentvalue.visible is true, this sets the prefix of the lable. If provided,',
+                'it will be joined to the current value with a single space between.'
+            ].join(' ')
+        },
```
@etpinard does the `prefix` definition look reasonable to you. I'm losing my mind here. The test dashboard fails:

```
> Plotly.PlotSchema.get().layout.layoutAttributes.sliders.currentvalue.prefix
< undefined
```

Where my browserifying dev environment succeeds:

```
> Plotly.PlotSchema.get().layout.layoutAttributes.sliders.currentvalue.prefix
< Object {valType: "string", role: "info", description: "When currentvalue.visible is true, this sets the p…to the current value with a single space between."}
```

I'm *really* struggling to find a reason for it. Currently digging into the logic to see if I've committed a crime which would get `prefix` filtered out *sometimes*…

rreusser replied:
Solved. Problem exists between screen, chair, and regex. (previous description lacked `.join()` 😬 )

rreusser commented:
@etpinard I've removed `updateevent` and `updatevalue` and cleaned up the slider options to more or less conform to axis options. The configurable options are now:

```
font
bgcolor
bordercolor
borderwidth
activebgcolor

ticklen
tickwidth
tickcolor
minorticklen

transition.duration
transition.easing

currentvalue.visible
currentvalue.xanchor
currentvalue.prefix
currentvalue.offset
currentvalue.font
```

The baseline is now within the range of what can be styled:

![sliders](https://cloud.githubusercontent.com/assets/572717/19155650/37209b8c-8bad-11e6-8b39-6dd8028be3b3.png)

etpinard commented on src/components/sliders/attributes.js:
```diff
@@ -209,22 +208,59 @@ module.exports = {
         description: 'Sets the width (in px) of the border enclosing the slider.'
-    }
+    },
+    ticklen: {
```
very nice

etpinard commented on src/components/sliders/defaults.js:
```diff
@@ -73,12 +73,17 @@ function sliderDefaults(sliderIn, sliderOut, layoutOut) {
-    Lib.coerceFont(coerce, 'currentvalue.font', layoutOut.font);
-
-    coerce('bgcolor', layoutOut.paper_bgcolor);
+    coerce('bgcolor');
```
I see  the default `bgcolor` isn't inherited from `layoutOut.paper_bgcolor` anymore. Is this on purpose?

rreusser replied:
To elaborate slightly, this attribute wasn't used before. Until today it was a hard-coded constant.

etpinard commented on src/components/sliders/attributes.js:
```diff
@@ -0,0 +1,266 @@
+
+        xanchor: {
+            valType: 'enumerated',
+            values: ['left', 'center', 'right'],
```
yep, nice job dropping the `'auto'` value here.

etpinard commented on src/components/sliders/attributes.js:
```diff
@@ -0,0 +1,266 @@
+            ].join(' ')
+        },
+
+        prefix: {
```
We should probably add a `suffix` attribute too.

rreusser replied:
✅

etpinard commented on src/components/sliders/defaults.js:
```diff
@@ -0,0 +1,116 @@
+    coerce('currentvalue.visible');
+    coerce('currentvalue.xanchor');
+    coerce('currentvalue.prefix');
+    coerce('currentvalue.offset');
```
this should be:

```js
var currentValue = coerce('currentvalue.visible');

if(currentValue) {
  coerce('currentvalue.xanchor');
  coerce('currentvalue.prefix');
  coerce('currentvalue.offset');
}
```

so that `Plotly.validate` can detect e.g. 

```js
layout.sliders = [{
  currentvalue: {
    visible: false,
    xanchor: 'left'
}];

// => `xanchor` has no effects
```

rreusser replied:
I've implemented this, but I don't fully understand it. Is this generally true that attributes shouldn't be coerced if the component is not visible?

rreusser replied:
✅

etpinard commented on src/components/sliders/defaults.js:
```diff
@@ -0,0 +1,116 @@
+    coerce('minorticklen');
+
+    Lib.coerceFont(coerce, 'font', layoutOut.font);
+    Lib.coerceFont(coerce, 'currentvalue.font', sliderOut.font);
```
similarly here, ` Lib.coerceFont(coerce, 'currentvalue.font', sliderOut.font);` should only get coerced if `currentvalue.visible` is true.

rreusser replied:
✅

etpinard commented on test/image/mocks/sliders.json:
```diff
@@ -0,0 +1,172 @@
+{
+  "data": [
```
great mock! Thanks :tada:

etpinard commented on test/jasmine/assets/fail_test.js:
```diff
@@ -23,4 +23,7 @@ module.exports = function failTest(error) {
         expect(error).toBeUndefined();
     }
+    if(error && error.stack) {
+        console.error(error.stack);
```
nicely done :tada:

etpinard commented on test/jasmine/tests/sliders_test.js:
```diff
@@ -0,0 +1,234 @@
+var destroyGraphDiv = require('../assets/destroy_graph_div');
+var fail = require('../assets/fail_test');
+
+describe('sliders defaults', function() {
```
TODO:

- [ ]  add one defaults around `currentvalue.visible` -> other `currentvalue` nested attributes

etpinard commented on test/jasmine/tests/sliders_test.js:
```diff
@@ -0,0 +1,234 @@
+    });
+});
+
+describe('update sliders interactions', function() {
```
Have you thought about adding mouse interaction tests?

It should be doable to mock a drag interaction using `new MouseEvent` - unless there are complications that I can't think about.

rreusser commented:
On a minor note, I precomputed the size of the text labels so that I could keep a right-anchored currentvalue prefix in the same place. The rationale is it makes it more clear which part is the value when written in the common use case as "label: value". I have mixed feelings on this.

![ezgif-2629780489](https://cloud.githubusercontent.com/assets/572717/19162567/5e11bcda-8bc6-11e6-886d-873c55978f75.gif)

rreusser commented:
@etpinard I've made the requested changes:

- [x] add suffix [70610ec](https://github.com/plotly/plotly.js/pull/986/commits/70610ec80a1ea8eb8ff85bd5da48093e97ffc016)
- [x] fix coerce when currentvalue !visible [f25127c](https://github.com/plotly/plotly.js/pull/986/commits/f25127cce9c2f556a0fb28376973b5f66bb1f1e1), [4b608ee](https://github.com/plotly/plotly.js/pull/986/commits/4b608eeb29e9c737f831491ffa7bda3b7987153c)
- [x] test mouse events (down, move, up) and associated appearance changes [6c9034d](https://github.com/plotly/plotly.js/pull/986/commits/6c9034d64fdbee21f42483fbfb338c1f4a749ec1)

etpinard commented on test/jasmine/tests/sliders_test.js:
```diff
@@ -249,6 +250,46 @@ describe('update sliders interactions', function() {
+            expect(mockCopy.layout.sliders[0].active).toEqual(0);
+
+            done();
+        }, 100);
```
Very nice. That's all I wanted. Thanks for taking the time to do this :tada:

rreusser replied:
No prob. Wasn't too bad; just hadn't previously tried mocking events and positions.

etpinard commented:
> I precomputed the size of the text labels so that I could keep the prefix in the same place for right-anchored currentvalue output. The rationale is it makes it more clear which part is the value with the common use case of "label: value".

I have no strong opinion about this. I'll you decide what's best.

rreusser commented:
I think it's fine as-is. I thought it would be preferable to completely right-justifying the text it since this cuts down on unneeded movement as you change the slider.

etpinard commented:
@rreusser I'll attach the gif here for reference:

![gifrecord_2016-10-06_151534](https://cloud.githubusercontent.com/assets/6675409/19167251/cda8a8e4-8bd8-11e6-9489-a10ac13bcc2a.gif)

looks like some race condition exist makes the `currentvalue` position change on relayout.

That happen once and I can't replicate it though.

So, let's merge this in :dancer: whenever you're happy with it.

rreusser commented:
Because it's particularly evasive (race condition?), @etpinard noted a layout issue that's captured here:

![gifrecord_2016-10-06_151534_360 gif png](https://cloud.githubusercontent.com/assets/572717/19167708/ecb582dc-8bda-11e6-95f9-410f4126e946.gif)

I think this was an issue with bounding box computation returning null and the sizes not actually being numbers. The last commit on this PR should fix that, but just adding here for reference in case it pops up again.
